### ISSUE TITLE
Remove .grep with .count > 1 Callables

### ISSUE THREAD

**Issue by zoffixznet** [rakudo/rakudo#1661]
This was first brought up as an example in https://github.com/rakudo/rakudo/issues/1657

`.grep` has an undocumented feature where you can give it a Callable with `.count > 2`:

```
<Zoffix__> m: dd (3, 1, 2).kv.grep: { $^a == $^b }
<camelia> rakudo-moar 637147af4: OUTPUT: «((1, 1), (2, 2)).Seq␤»
```

This behavious is baffling to me, as `.grep` is supposed to be a means to **filter** the data, yet in this special case it also mutates it. 

The way it behave at the moment isn't how I would've expected it to consume the original data, so there's also an issue of producing some arcane, unintuitive behaviour.

The alternative for the same output is rather straightforward and is only 4-characters longer:

```
<Zoffix__> m: dd (3, 1, 2).kv.rotor(2).grep: { [==] $_  }
<camelia> rakudo-moar 637147af4: OUTPUT: «((1, 1), (2, 2)).Seq␤»
```

In addition, there's a `.map` alternative as well:

```
<Zoffix__> m: dd (3, 1, 2).kv.map: { $^a == $^b or next; ($a, $b) }
<camelia> rakudo-moar 637147af4: OUTPUT: «((1, 1), (2, 2)).Seq␤»
```

Lastly, I see exactly one test in 6.c roast for this behaviour:

```
my @in = ( 1, 1, 2, 3, 4, 4 );
is (grep { $^a == $^b }, @in), (1, 1, 4, 4), 'grep takes two at a time';
```

You may notice the resultant output isn't actually structured to how the author expected the output to
be produced. The reason the test passes at all is because of inappropriate test function `is` being used. `is` stringifies its arguments and destroys all structure.

----------

To summarize, my reasons for removal of this feature are:

* It's a special case of .grep that mutates original stream of data by `.rotor`ing it. No other arguments do this.
* It's confusing behaviour, as the mutation of original data happens behind the scenes
* It's easy to replicate this uncommonly-wanted behaviour with explicit `.rotor` or a map call.
* The only spec test covering this behaviour is passing for the wrong reason and the current behaviour doesn't match what the original designer expected this feature to behave as.

**Comment by zoffixznet** [comment]
And 1 more argument for its removal. It's currently broken for all but `:v` versions of grep. So seems it was added to roast as an afterthought, pre-GLR and the roasted test is passing for the wrong reason.

```
<Zoffix__> m: <a b c>.kv.grep(-> $, $ { True }, :kv).say
<camelia> rakudo-moar e31dd966f: OUTPUT: «Too few positionals passed; expected 2 arguments but got 1␤  in block <unit> at <tmp> line 1␤␤»
<Zoffix__> m: <a b c>.kv.grep(-> $, $ { True }, :k).say
<camelia> rakudo-moar e31dd966f: OUTPUT: «Too few positionals passed; expected 2 arguments but got 1␤  in block <unit> at <tmp> line 1␤␤»
<Zoffix__> m: <a b c>.kv.grep(-> $, $ { True }, :p).say
<camelia> rakudo-moar e31dd966f: OUTPUT: «Too few positionals passed; expected 2 arguments but got 1␤  in block <unit> at <tmp> line 1␤␤»
<Zoffix__> m: <a b c>.kv.grep(-> $, $ { True }, :v).say
<camelia> rakudo-moar e31dd966f: OUTPUT: «((0 a) (1 b) (2 c))␤»
```

**Comment by AlexDaniel** [comment]
> It's easy to replicate this uncommonly-wanted behaviour with explicit .rotor or a map call.

It's different a bit different though:
```perl6
say <a b c d e>.rotor(2).grep: { True } 

((a b) (c d))
```
```perl6
say <a b c d e>.grep: -> $a, $b { True }

Too few positionals passed; expected 2 arguments but got 1
  in block <unit> at -e line 1
```

FWIW `map` can also take a callable with .count > 1, so the fact that it works isn't very surprising to me… but I do agree that grep that does some `mapping` is weird. Very weird. And it's so visually unclear that I'll never write it like this…

I'm a bit conflicted because it feels like it should do something sensible, but the reasoning here is about right… so +½ on deprecation and further removal. I guess that was the plan, right?

**Comment by AlexDaniel** [comment]
Er, scratch the first part of my comment. Obviously you can use .map and get the identical behavior if you need it.

**Comment by zoffixznet** [comment]
> deprecation and further removal. I guess that was the plan, right?

Right. Warn in 6.d and remove in 6.e

**Comment by AlexDaniel** [comment]
Sounds good to me. In fact, it's maybe bit too forgiving for an undocumented and undertested feature that doesn't even work fully. But yeah, someone somewhere is probably using that…

**Comment by AlexDaniel** [comment]
Actually… my other +½ was for turning it into some feature that makes sense and doesn't do weird stuff, but now I'm realizing that if we want to *change* the behavior we'll have to go through the deprecation process anyway… so I'm actually full +1 on removing it like planned. If someone ever comes up with a well-thought-out RFC, then we can adjust the direction later.
